src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj

```diff
@@ -356,34 +339,137 @@
 
 (schema/defn ^:always-validate environment-class-info-entry
   :- EnvironmentClassInfoCacheEntry
-  "Create an environment class info entry"
-  ([]
-   (environment-class-info-entry nil))
-  ([tag :- (schema/maybe schema/Str)]
-   {:tag tag
-    :last-updated (System/currentTimeMillis)}))
+  "Create an environment class info entry.  The return value will have a nil
+  tag and cache-generation-id set to 1."
+  []
+  {:tag nil
+   :cache-generation-id 1})
+
+(schema/defn ^:always-validate inc-cache-generation-id-for-class-info-entry
+  :- EnvironmentClassInfoCacheEntry
+  "Return the supplied 'original-environment-class-info-entry' class info entry,
+  only updated with a cache-generation-id value that has been incremented.
+  If the supplied parameter is nil, the return value will have a tag set to
+  nil and cache-generation-id set to 1."
+  [original-environment-class-info-entry :-
+   (schema/maybe EnvironmentClassInfoCacheEntry)]
+  (if original-environment-class-info-entry
+    (update original-environment-class-info-entry :cache-generation-id inc)
+    (environment-class-info-entry)))
+
+(schema/defn ^:always-validate update-environment-class-info-entry
+  :- EnvironmentClassInfoCacheEntry
+  "Return the supplied 'original-environment-class-info-entry', only updated
+  with the supplied tag and a cache-generation-id value that has been
+  incremented."
+  [original-environment-class-info-entry :-
+   (schema/maybe EnvironmentClassInfoCacheEntry)
+   tag :- (schema/maybe schema/Str)]
+  (-> original-environment-class-info-entry
+      inc-cache-generation-id-for-class-info-entry
+      (assoc :tag tag)))
+
+(schema/defn ^:always-valid invalidated-environment-class-info-entry
```

Comment:
Nitpick that I should have commented on this before you updated the PR, but this function name also has `invalidated` rather than `invalidate`.